warehouse/query-core/src/main/java/datawave/query/util/transformer/AttributeRebuilder.java

```diff
@@ -0,0 +1,124 @@
+package datawave.query.util.transformer;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import javax.annotation.Nullable;
+
+import org.apache.log4j.Logger;
+
+import datawave.data.type.Type;
+import datawave.query.attributes.Attribute;
+import datawave.query.attributes.TypeAttribute;
+import datawave.query.jexl.JexlASTHelper;
+import datawave.query.model.QueryModel;
+import datawave.query.util.TypeMetadata;
+
+/**
+ * Utility class that handles rebuilding a basic NoOp attribute into an attribute of the correct type, given TypeMetadata and optionally a QueryModel
+ */
+public class AttributeRebuilder {
+    private static final Logger log = Logger.getLogger(AttributeRebuilder.class);
+    private final TypeMetadata typeMetadata;
+    private final Map<String,String> fieldMap;
+    private final Map<String,Class<?>> classCache;
+
+    /**
+     * Main constructor for the rebuilder, QueryModel may be null
+     *
+     * @param typeMetadata
+     * @param queryModel
+     */
+    public AttributeRebuilder(TypeMetadata typeMetadata, @Nullable QueryModel queryModel) {
+        this.typeMetadata = typeMetadata;
+        if (queryModel == null) {
+            this.fieldMap = new HashMap<>();
+        } else {
+            this.fieldMap = invertMap(queryModel.getReverseQueryMapping());
+        }
+        this.classCache = new HashMap<>();
+    }
+
+    private Map<String,String> invertMap(Map<String,String> map) {
+        Map<String,String> mappings = new HashMap<>();
```

Comment:
This may need to be a Multimap.  There is nothing preventing us from having multiple fields with the same reverse mapping, although I don't believe we do this in practive.